cheatsheets/Microservices_Security_Cheat_Sheet.md

```diff
@@ -2,176 +2,660 @@
+
+* **Operational complexity:** Issuing, validating, and revoking tokens requires careful coordination and infrastructure to ensure security and scalability.
+* **Issuer scalability and reliability:** The token service must be highly available and performant to avoid bottlenecks or single points of failure.
+* **Dependency on transport-layer security:** Token-based authentication requires TLS to protect token confidentiality and prevent replay attacks.
+
+
+## Authorization Patterns
+
+Given the reference architecture described above, we can now examine common authorization patterns and explore the trade-offs they entail.
+
+### Decentralized Service-Level Access Control
+
+In this pattern, most of the functional components from the reference architecture are implemented directly within each microservice. Even the Policy Information Points (PIPs) may be embedded into the service logic (e.g., via database or configuration entries) if the microservice is responsible for the all relevant attributes itself. However, this is rarely the case, and most microservices must integrate with other services to retrieve required attributes, treating those other services as external PIPs.
+
+![Decentralized Service-Level Access Control](../assets/Decentralized_Service_Level_Access_Control.svg)
+
+The access control rules are typically implemented using native language constructs (e.g., `if`/`else` statements), either inline with business logic functions or via abstraction mechanisms such as interceptors.
+
+When a microservice receives a request containing authorization metadata (e.g., end-user context or resource identifiers), it evaluates whether access should be granted. This may involve querying other services (PIPs) for additional attributes before reaching a decision and enforcing it (implicitly). Alternatively, some services may use asynchronous communication patterns (e.g., periodic syncs or event-driven updates) to pre-fetch required data in advance, improving performance and resilience.
+
+When adopting this approach, the following trade-offs should be considered:
+
+#### Pros
+
+* **Familiar development model**: Developers can use the same language and tools they already know.
+* **Framework support**: Many libraries and frameworks exist for many languages to reduce boilerplate and simplify integration.
+* **Rapid prototyping**: Policy logic is implemented directly in code, enabling quick experimentation and iteration.
+* **Team autonomy**: Fits well with domain-driven design and independent team ownership; each team can choose its approach.
+* **High performance**: Policy evaluation is done in-memory within the microservice.
+* **Full context awareness**: The service has access to runtime data, business logic, and domain models, enabling fine-grained, context-rich and nuanced decisions.
+* **Failure isolation**: If all required attributes are available locally or cached, failures in external systems do not impact decision-making.
+
+#### Cons
```

Comment:
(possibly) missing Con:

**Intermingling of business- and access-related code**: Access requirements and Business requirements often have distinct lifecycles. Adding authorization-related code to the services codebase may result in the need to redeploy the service when there are changes related to access (e.g., a subscription-based service needs a bigger rewrite because now the "gold" subscription covers less than before)

if that makes sense.